### PR TITLE
Add MySQL support to DataStore SQL plugin

### LINKED ISSUE #726: Mysql support for SPIRE Server
We plan to deploy SPIRE in our production environments.  Our preferred database & the one we have operational tooling for is Mysql.

Thus we'd like support for that database.  

It seems there was an earlier attempt at supporting it but it appears abandoned/unmerged:
https://github.com/spiffe/spire/pull/530

What is needed to support Mysql?

### PR DESCRIPTION
**Pull request by mbyczkowski**
This change is based off of original work by @amonetta in
https://github.com/spiffe/spire/pull/530

Signed-off-by: Mat Byczkowski <mbyczkowski@squareup.com>

**Pull Request check list**

- [x] Commit conforms to CONTRIBUTING.md?
- [x] Proper tests/regressions included?
- [x] Documentation updated?

**Affected functionality**
Adds MySQL support to DataStore SQL plugin

**Description of change**
Most of the PR is just sample config files, the rest is:
- connection init for MySQL
- tweaks to migration code that allows MySQL-specific options to be set when creating new tables
- explicit size set on the byte array, so appropriate type is selected for the underlying column in MySQL
- integration tests for MySQL

I have performed a bunch of ad-hoc tests and this is the minimal set of changes needed to get MySQL support. There are some gotchas and rough edges, most of them around index size for `VARCHAR` columns.

By default MySQL has had a hard limit of 767 bytes for the index (which most people learned about through this [infamous error](https://stackoverflow.com/questions/1814532/1071-specified-key-was-too-long-max-key-length-is-767-bytes)). Later the limit was raised to 3072 (the `innodb_large_prefix` option that was set as a default in 5.7.7). Therefore, I have explicitly set the default character set for each table to `utf8` (and not the full UTF-8, which in MySQL is called `utf8mb4`), so that regardless of the MySQL server config we can support current schema.

This brings me to the last point, current mapping of `string` fields in SPIRE models are mapped to `VARCHAR(255)` in MySQL, which might not be enough (at least in comparison to Sqlite and PostgreSQL). If that's the case, it would make sense to add more specific limits with `size:` GORM tags. However, then we might not be able to support unique indexes on those fields, and would have to create new model structs specific to MySQL.

**Which issue this PR fixes**
fixes #726

### CODE REVIEW

File: conf/server/server-postgres.conf [hunk 1]
```diff
@@ -0,0 +1,54 @@
+server {
```
Comment (mbyczkowski) [inline conf/server/server-postgres.conf]: I found this config helpful for running tests with local PostgreSQL, but I'm happy to remove it.

File: conf/server/server-mysql.conf [hunk 2]
```diff
@@ -0,0 +1,55 @@
+server {
```
Comment (mbyczkowski) [inline conf/server/server-mysql.conf]: This could be moved to `ha-mysql.conf` to mirror current PostgreSQL config

File: pkg/server/plugin/datastore/sql/migration.go [hunk 3]
```diff
@@ -92,6 +100,16 @@ func initDB(db *gorm.DB) (err error) {
 	return nil
 }
 
+func tableOptionsForDialect(tx *gorm.DB, dbType string) *gorm.DB {
+	// This allows for setting table options for a particular DB type.
+	// For MySQL, (for compatibility reasons) we want to make sure that
+	// we can support indexes on strings (varchar(255) in the DB).
+	if dbType == "mysql" {
+		return tx.Set("gorm:table_options", "ENGINE=InnoDB  ROW_FORMAT=DYNAMIC DEFAULT CHARSET=utf8")
```
Comment (dweitzman) [inline pkg/server/plugin/datastore/sql/migration.go]: Random drive-by minor comment: seems like the charset named "utf8mb4" would always be a better choice in mysql than the charset named "utf8", since in mysql "utf8mb4" means "standard utf8", and "utf8" means "a character set invented by the mysql team that's a subset of utf8"

Comment (dweitzman) [reply pkg/server/plugin/datastore/sql/migration.go]: Oh, I see that you had a comment about the index error. Ignore me. This compromise probably won't hurt anyone :)

File: conf/server/server-mysql.conf [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+mysql> CREATE USER 'spire'@'localhost';
+mysql> GRANT ALL PRIVILEGES ON spire_development.* TO 'spire'@'localhost'
+mysql> FLUSH PRIVILEGES;
+```
+*/
+plugins {
+    DataStore "sql" {
+        plugin_data {
+            database_type = "mysql"
+            connection_string = "spire:@tcp(127.0.0.1)/spire_development?parseTime=true"
```
Comment (azdagron) [inline conf/server/server-mysql.conf]: From what I can tell, `parseTime` is required for the plugin to operate properly. Configure() should probably fail if `parseTime` isn't set to `true`, or alternatively parse the connection string and unconditionally add it if missing.

File: pkg/server/plugin/datastore/sql/models.go [hunk 5]
```diff
@@ -16,7 +17,7 @@ type Bundle struct {
 	Model
 
 	TrustDomain string `gorm:"not null;unique_index"`
-	Data        []byte
+	Data        []byte `gorm:"size:65535"`
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/models.go]: Is this a required constraint for MySQL? How did we arrive at 64K? The upper bound on this field is a function of how many SPIRE servers are running and the CA TTL.... I this there is potential to blow this limit.

Also, how will this impact other non-MySQL databases...?

File: script/e2e_test.sh [hunk 6]
```diff
@@ -85,4 +85,8 @@ run_test() {
 }
 
 run_e2e_test "conf/server/server.conf"
-run_docker_test "test/configs/server/postgres.conf" "-e POSTGRES_PASSWORD=password -p 5432:5432 -d postgres"
+run_docker_test "test/configs/server/postgres.conf" "-e POSTGRES_PASSWORD=password -p 10864:5432 -d postgres"
+run_docker_test "test/configs/server/mysql.conf" "-e MYSQL_PASSWORD=password -e MYSQL_DATABASE=mysql -e MYSQL_USER=mysql -e MYSQL_RANDOM_ROOT_PASSWORD=yes -p 6612:3306 -d mysql:5.5.62"
```
Comment (azdagron) [inline script/e2e_test.sh]: hmm, I wonder if we should be running all these tests against different versions on every CI run? It may be sufficient to only run against the most prevalent version and run the others on merge to master and/or tagged builds (same criteria we use to build images and run more lengthy systems tests)...

File: conf/server/server-mysql.conf (hunk 4 above)
Comment (mbyczkowski) [reply conf/server/server-mysql.conf]: Agreed. I can add a validation for that.

File: script/e2e_test.sh (hunk 6 above)
Comment (mbyczkowski) [reply script/e2e_test.sh]: Yes, I didn't realize that certain tests were only running on merges/tagged branches. I think this should go there and we could also add different versions of Postgres to those tests (later as separate PR).

File: pkg/server/plugin/datastore/sql/models.go (hunk 5 above)
Comment (mbyczkowski) [reply pkg/server/plugin/datastore/sql/models.go]: Without specifying the size, `data` column in MySQL was of type `VARBINARY(255)`. I picked the upper limit of `BLOB` and indeed it allows up to 64kB of data - it could have been any number > 255, and I figured that having this particular value would be good enough. We can set it to `MEDIUMBLOB` though - so anything between 65535 and 16777215 (24MB).

Regardless of this change, when I tried to initialize PostgreSQL (both on master and this branch), the `data` column was of type `BYTEA`. Similarly, Sqlite was setting `BLOB` in both cases, so no negative impact observed.

Comment (azdagron) [reply pkg/server/plugin/datastore/sql/models.go]: I think I'd be a little more keen on a 24MB upper boundary. Unfortunately the failure mode in this scenario could be pretty terrible but it is what it is. We can revisit things at a later time if people start bumping up against this.

File: pkg/server/plugin/datastore/sql/sql.go [hunk 7]
```diff
@@ -39,6 +39,12 @@ var (
 	sqlError = errs.Class("datastore-sql")
 )
 
+const (
+	MySQL      = "mysql"
+	PostgreSQL = "postgresql"
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/sql.go]: This was previously "posgres".

Comment (mbyczkowski) [reply pkg/server/plugin/datastore/sql/sql.go]: Yes, bad refactoring. Going to revert back to `postgres` 👍

File: pkg/server/plugin/datastore/sql/sql.go [hunk 8]
```diff
@@ -1360,3 +1365,22 @@ func bindVarsFn(fn func(int) string, query string) string {
...
+func validateDBConfig(cfg *configuration) error {
+	if cfg.DatabaseType == "" {
+		return errors.New("database_type must be set")
+	}
+
+	if cfg.ConnectionString == "" {
+		return errors.New("connection_string must be set")
+	}
+
+	switch cfg.DatabaseType {
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/sql.go]: nit: we don't need this switch statement if each dialect takes care of validating the connection string inside of the various connect() methods.
